File: neetcode/python/875.koko-eating-bananas.py
#
# @lc app=leetcode id=875 lang=python3
#
# [875] Koko Eating Bananas
#

# @lc code=start
class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        
        def can_finish(piles, mid):
            time_taken = 0

            for num_bananas in piles:
                time_taken += math.ceil(num_bananas / mid)
            return time_taken <= h    

        high = max(piles)
        low = 1

        while low < high:
            mid = low + (high - low) // 2
            if can_finish(piles, mid):
                high = mid
            else:
                low = mid + 1
        return high
# @lc code=end

File: neetcode/go/875.koko-eating-bananas.go
/*
 * @lc app=leetcode id=875 lang=golang
 *
 * [875] Koko Eating Bananas
 */

// @lc code=start
func minEatingSpeed(piles []int, h int) int {
	CanFinish := func(speed int) bool {
		timeTaken := 0
		for _, numBananas := range piles {
			if numBananas%speed != 0 {
				timeTaken += (numBananas / speed) + 1
			} else {
				timeTaken += numBananas / speed
			}
		}
		return timeTaken <= h
	}

	low := 1
	high := slices.Max(piles)

	for low < high {
		mid := low + (high-low)/2
		if CanFinish(mid) {
			high = mid
		} else {
			low = mid + 1
		}
	}

	return high
}

// @lc code=end

